crates/sui-framework/sources/governance/staking_pool.move

```diff
@@ -0,0 +1,343 @@
+        /// field will record the original lock expiration epoch, to be used when unstaking.
+        sui_token_lock: Option<EpochTimeLock>,
+    }
+
+    /// Create a new, empty staking pool.
+    public(friend) fun new(validator_address: address, starting_epoch: u64) : StakingPool {
+        StakingPool {
+            validator_address,
+            starting_epoch,
+            epoch_starting_sui_balance: 0,
+            sui_balance: 0,
+            rewards_pool: balance::zero(),
+            delegation_token_supply: balance::create_supply(DelegationToken {}),
+            pending_delegations: vector::empty(),
+        }
+    }
+
+    /// Called at epoch advancement times to add rewards (in SUI) to the staking pool, and distribute new delegation tokens. 
+    public(friend) fun advance_epoch(pool: &mut StakingPool, rewards: Balance<SUI>, ctx: &mut TxContext) {
+        pool.sui_balance = pool.sui_balance + balance::value(&rewards);
+        balance::join(&mut pool.rewards_pool, rewards);
+        
+        // distribute pool tokens at new exchange rate.
+        while (!vector::is_empty(&pool.pending_delegations)) {
+            let PendingDelegationEntry { delegator, sui_amount } = vector::pop_back(&mut pool.pending_delegations);
+            distribute_delegation_tokens(pool, delegator, sui_amount, ctx);
+            pool.sui_balance = pool.sui_balance + sui_amount
+        };
+
+        // Record the epoch starting balance.
+        pool.epoch_starting_sui_balance = pool.sui_balance;
+    }
+
+    // TODO: implement rate limiting new delegations per epoch.
+    /// Request to delegate to a staking pool. The delegation gets counted at the beginning of the next epoch,
+    /// when the delegation object containing the pool tokens is distributed to the delegator.
+    public(friend) fun request_add_delegation(
+        pool: &mut StakingPool, 
+        stake: Balance<SUI>, 
+        sui_token_lock: Option<EpochTimeLock>, 
+        ctx: &mut TxContext
+    ) {
+        let sui_amount = balance::value(&stake);
+        assert!(sui_amount > 0, 0);
+        let delegator = tx_context::sender(ctx);
+        // insert delegation info into the pendng_delegations vector.         
+        vector::push_back(&mut pool.pending_delegations, PendingDelegationEntry { delegator, sui_amount });
+        let staked_sui = StakedSui {
+            id: object::new(ctx),
+            principal: stake,
+            sui_token_lock,
+        };
+        transfer::transfer(staked_sui, delegator);
+    }
+
+    /// Activate a delegation. New pool tokens are minted at the current exchange rate and put into the
+    /// `pool_tokens` field of the delegation object.
+    /// After activation, the delegation officially counts toward the staking power of the validator.
+    /// Aborts if the pool mismatches, the delegation is already activated, or the delegation cannot be activated yet. 
+    public(friend) fun distribute_delegation_tokens(
```

Comment:
"distribute" usually has the meaning of giving things to a list of entities. Should probably use a different name here.